### PR TITLE
[WIP] some mir typeck cleanup

### PR DESCRIPTION
**Pull request by lcnr**
TODO: write this

r? @nikomatsakis

### CODE REVIEW

File: compiler/rustc_borrowck/src/type_check/mod.rs [hunk 1]
```diff
@@ -511,6 +511,15 @@ impl<'a, 'b, 'tcx> TypeVerifier<'a, 'b, 'tcx> {
             place_ty = self.sanitize_projection(place_ty, elem, place, location);
         }
 
+        self.cx.prove_predicate(
+            ty::Binder::dummy(ty::PredicateKind::WellFormed(
+                self.body().local_decls[place.local].ty.into(),
+            ))
+            .to_predicate(self.tcx()),
+            location.to_locations(),
+            ConstraintCategory::Boring,
```
Comment (nikomatsakis) [inline compiler/rustc_borrowck/src/type_check/mod.rs:518]: do we have a known bug that is affected by this?

File: compiler/rustc_borrowck/src/type_check/mod.rs [hunk 2]
```diff
@@ -400,14 +400,12 @@ impl<'a, 'b, 'tcx> Visitor<'tcx> for TypeVerifier<'a, 'b, 'tcx> {
...
-                    instantiated_predicates,
-                    location.to_locations(),
-                );
-            }
+            self.cx.prove_predicate(
+                ty::Binder::dummy(ty::PredicateKind::WellFormed(constant.literal.ty().into()))
+                    .to_predicate(self.tcx()),
+                location.to_locations(),
+                ConstraintCategory::Boring,
+            );
```
Comment (nikomatsakis) [inline compiler/rustc_borrowck/src/type_check/mod.rs:408]: didn't we file a bug about this?

**Review by nikomatsakis — APPROVED** [review]
looks good to me so far! is this still a WIP?

**Comment by pnkfelix** [comment]
lcnr reports that this still has a nasty bug that they need to work out.

@rustbot label: +S-waiting-on-author -S-waiting-on-review

**Comment by lcnr** [comment]
closing in favor of #100761, I intend to work on the the refactorings I've dropped at a later date

**Pull request closed without merging**